docs/node-mixin/lib/common.libsonnet

```diff
@@ -0,0 +1,714 @@
+local grafana = import 'github.com/grafana/grafonnet-lib/grafonnet/grafana.libsonnet';
+local dashboard = grafana.dashboard;
+local row = grafana.row;
+local prometheus = grafana.prometheus;
+local template = grafana.template;
+local nodePanels = import '../lib/panels/panels.libsonnet';
+local commonPanels = import '../lib/panels/common/panels.libsonnet';
+local nodeTimeseries = nodePanels.timeseries;
+{
+
+  new(config=null, platform=null):: {
+
+    local c = self,
+
+    local labelsToRegexSelector(labels) =
+      std.join(',', ['%s=~"$%s"' % [label, label] for label in labels]),
+    local labelsToLegend(labels) =
+      std.join('/', ['{{%s}}' % [label] for label in labels]),
+
+    local labelsToURLvars(labels, prefix) =
+      std.join('&', ['var-%s=${%s%s}' % [label, prefix, label] for label in labels]),
+    // export
+    labelsToLegend:: labelsToLegend,
+    labelsToURLvars:: labelsToURLvars,
+    // add to all queries but not templates
+    local nodeQuerySelector = labelsToRegexSelector(std.split(config.groupLabels + ',' + config.instanceLabels, ',')),
+    nodeQuerySelector:: nodeQuerySelector,
+
+    // common templates
+    local prometheusDatasourceTemplate = {
+      current: {
+        text: 'default',
+        value: 'default',
+      },
+      hide: 0,
+      label: 'Data Source',
+      name: 'datasource',
+      options: [],
+      query: 'prometheus',
+      refresh: 1,
+      regex: '',
+      type: 'datasource',
+    },
+
+    local chainLabelsfold(prev, label) = {
+      chain:
+        if std.length(prev) > 0
+        then
+          [[label] + prev.chain[0]] + prev.chain
+        else
+          [[label]],
+    },
+
+    local chainLabels(labels) =
+      [
+        {
+          label: l[0:1][0],
+          chainSelector: labelsToRegexSelector(std.reverse(l[1:])),
+        }
+        for l in std.reverse(std.foldl(chainLabelsfold, labels, init={}).chain)
+      ],
+
+    local groupTemplates =
+      [
+        template.new(
+          name=label.label,
+          label=label.label,
+          datasource='$datasource',
+          query='',
+          current='',
+          refresh=2,
+          includeAll=true,
+          // do not use .* will get series without such label at all when ALL is selected.
+          // do not use .+ will ignore nodeExporterSelector results
+          // use null for group values
+          allValues=null,
```

Comment:
Is `.+` the default value for grafonnet if `allValues` is null? Should `allValues` be explicitly set to `.+`?

Reply:
no there is no custom 'allValues' in here on purpose:

```
          // do not use .* will get series without such label at all when ALL is selected.
          // do not use .+ will ignore nodeExporterSelector results
```
1. There is a nodeExporterSelector (filter) that intends to narrow down the scope of node_exporter mixin (like macos/raspberry pi/kubernetes or other mixins that maybe based on metrics of node_exporter do not interfere). It used in dashboard templates queries only.
2. There is a nodeQuerySelector that is put in every dashboard query (which is basically group+instance labels selector, and actualy  `job=~$job, instance=~$instances by default).

If we use .+ for job allValue, then it will ignore nodeExporterSelector filter applied and spill over.